### DISCUSSION TITLE
Checkboxes to control the display of multiple mesh

### DISCUSSION THREAD

**Discussion by Oguitton** [Kitware/trame#268] [Q&A] [answered]
Hi,

I apologize in advance for my bad English I am French.

I try to show some electrical tomographic profiles and to control the display of it.
I would like to code checkboxes with the name of the profile on the left with a SinglePageWithDrawerLayout().
For now, I achieved to code VBtn : one to show and one to hide the profile with the SetVisibility() function. But when I try with the VCheckbox it does not work, I don't know why. In the code, I say what I want it to do for the "true_value" and the "false-value" but I think it is not the right way.

Another question is about an iteration. 
When I have many profils I would like to code a "for" loop to create the various checkboxes. But because it is dynamic, it does not work with an "i" because this is a variable. 

Here is an extract from the code : 
with file_tot a list with all the name's profile I want to show and the Vbtn I would like to replace by VCheckboxes.

with SinglePageWithDrawerLayout(server) as layout:
    
    layout.title.set_text(state.trame__title)
    create_mesh(actor)
    with layout.drawer:
        vuetify.VSpacer()                 
        vuetify.VCheckbox(label = file_tot[0], true_value = remove_profil(0,1)), false_value = 	remove_profil(0,0))
        
        vuetify.VBtn(name_mesh[0], click=lambda: remove_profil(0,1))#,style="position: absolute; 	top: 100; right: 50; z-index: 10;",)
        vuetify.VBtn("hide", click=lambda: remove_profil(0,0))
        vuetify.VBtn(name_mesh[1], click=lambda: remove_profil(1,1)) 
        vuetify.VBtn("hide", click=lambda: remove_profil(1,0))
        
        
    with layout.content:
        with vuetify.VContainer(
        
            fluid=True,
            classes="pa-0 fill-height",
        ):
            
            # Use PyVista UI template for Plotters
            view = plotter_ui(pl)
            ctrl.view_update = view.update

server.start()

and the remove_profil's function : 

#-------------fonction setvisibility -------
#avec la valeur de flag la fonction affiche ou cache le mesh
def remove_profil(num,flag):
    # pl.remove_actor(actor[num])
    # pl.update()
    # pl.show()
    actor[num].SetVisibility(flag)
    pl.update()
    pl.show()

Thanks a lot for your help !

**Comment by jourdain** [comment, marked as answer]
The checkbox documentation is available [here](https://v2.vuetifyjs.com/en/api/v-checkbox/) but in general you want to use a `v_model` to control your variable. The true/false_value is to remap the true/false to something else for your variable such as "show", "hide" string if preferred. 

So the basic gist would be

```python
with layout:
   # ...
   vuetify.VCheckbox(v_model=("visible", True))

And then

@state.change("visible")
def on_visible_change(visible, **kwargs):
    if visible:
        show()
    else:
        hide()
``` 

Regarding your dynamic list question, you  can look at https://github.com/Kitware/trame/discussions/253 for reference

But for you it could be something like

```python
state.profileVisibility = [True, True, True, True, True]

with layout:
   # ...
   vuetify.VCheckBox(
       v_for="v, i in profileVisibility",
       key="i",
       label=("Toggle visibility of {{ i }}",),
       value=("v",),
       change=(update_profile, "[i, $event]"),
   )

def update_profile(index, visible):
    state.profileVisibility[index] = visible
    state.dirty("profileVisibility")
    # do your show/hide
```

**Comment by Oguitton** [comment]
Thanks a lot for your help ! 
The v_model parameter is more understanble and it worked with one checkbox.
But I struggle a little bit with the dynamic list. I try something with the v_for and a label and the v_model. But I don't konw how to access to the "i" variable which I need for the name because it is on a tab, and for the on_visible_change function. 

```python
vuetify.VCheckbox(v_for="v, i in profileVisibility",key="i", label = 'essai', v_model=("visible", True))
I use this line for the checkbox, and the function are : 

@state.change("visible")
def on_visible_change(visible, **kwargs):
    if visible:
        remove_profil(0,True)
    else:
        remove_profil(0,False)


def remove_profil(num,flag):
    actor[num].SetVisibility(flag)
    ctrl.view_update()

```

For now, I fixed the "num" variable to "0" on the remove_profil function and it worked, I have several checkbox with the same name "essai" and all of them control a unique mesh. 
And I try your second code but I think the line : label =("Toggle visibility of {{ i }}",) induce a probleme because without it it work and with it nothing hapen on the web page.

**Comment by jourdain** [comment]
So to expend my example to fit your need, you just need to edit the function below

```python
def update_profile(index, visible):
    state.profileVisibility[index] = visible
    state.dirty("profileVisibility")
    # do your show/hide
    actor[index].SetVisibility(visible)
    ctrl.view_update()
```

**Reply by jourdain** [reply]
No more `@state.change(...)`, but a single event listener instead. Basically you can not rely on a boolean variable to capture the state of several checkboxes. You need a list instead.

**Reply by Oguitton** [reply]
Yes I see I did not understand the comment "do your show/hide" before, now I see. 
What is the list you talk about ? Is it the profilVisibility you define in the first answer ? 
I don't now why but the 'i' index did not work, and with the label I still have the same problem : a blank page without anything including the toolbar, really nothing. 
Thanks a lot for your time and your help !

**Reply by jourdain** [reply]
Good point small typo on the label templating... Fully working example below

```python
from trame.app import get_server
from trame.widgets import vuetify
from trame.ui.vuetify import SinglePageLayout

server = get_server()
state, ctrl = server.state, server.controller

state.visibilityList = [True, False, True, False]

def update_visibility(index, visibility):
    state.visibilityList[index] = visibility
    state.dirty("visibilityList")
    print(f"Toggle {index} to {visibility}")


with SinglePageLayout(server) as layout:
    with layout.content:
        with vuetify.VContainer(fluid=True, classes="fill-height pa-0 ma-0"):
            with vuetify.VCol():
                vuetify.VCheckbox(
                    v_for="v, i in visibilityList",
                    key="i",
                    label=("`Hello ${i}`",),
                    v_model=("visibilityList[i]",),
                    change=(update_visibility, "[i, $event]"),
                )


if __name__ == "__main__":
    server.start()
```

**Reply by Oguitton** [reply]
thanks a lot for your help it work very well ! 
I just add my remove_profil function in your update_visibility and it is all. 
Thanks again and have a nice day. 
Oriane
